vault-plugins/flant_iam_auth/usecase/authn/accesstoken/validator.go

```diff
@@ -0,0 +1,319 @@
+		return nil, errors.New("invalid expiration time (exp) claim: token is expired")
+	}
+	if claims.IssuedAt != nil && now.Add(cksLeeway).Before(claims.IssuedAt.Time()) {
+		return nil, errors.New("invalid issued at (iat) claim: token issued in the future")
+	}
+
+	return allClaims, nil
+}
+
+// enrichedClaims validates access_token, and returns enriched claims
+//
+// The given JWT is considered valid if:
+//  0. iss match issuer at OIDCDiscoveryURL
+//  1. oidc-provider returns userinfo presenting access_token
+//
+//  enrichedClaims returns claims given by the token and new claims from userinfoURL
+func (v *accessTokenValidator) enrichedClaims(ctx context.Context, token string) (map[string]interface{}, error) {
+	userinfo, err := v.getUserInfo(ctx, token)
+
+	parsedJWT, err := jwt.ParseSigned(token)
+	if err != nil {
+		return nil, err
+	}
+	allClaims := map[string]interface{}{}
+	err = parsedJWT.UnsafeClaimsWithoutVerification(&allClaims)
+	if err != nil {
+		return nil, err
+	}
+	for k, v := range userinfo {
+		if _, allredyExists := allClaims[k]; !allredyExists {
+			allClaims[k] = v
+		}
+	}
+	return allClaims, nil
+}
+
+func (v *accessTokenValidator) getUserInfo(ctx context.Context, token string) (map[string]interface{}, error) {
+	req, err := http.NewRequest(http.MethodGet, v.userinfoURL, nil)
+	if err != nil {
+		return nil, err
+	}
+	req.Header.Add("Authorization", "Bearer "+token)
+	resp, err := v.client.Do(req.WithContext(ctx))
+	if err != nil {
+		return nil, err
+	}
+	defer resp.Body.Close()
+
+	body, err := ioutil.ReadAll(resp.Body)
+	if err != nil {
+		return nil, fmt.Errorf("unable to read response body: %v", err)
+	}
+	if resp.StatusCode != http.StatusOK {
+		return nil, fmt.Errorf("%s: %s", resp.Status, body)
+	}
+
+	userinfo := map[string]interface{}{}
+	if err := unmarshalResp(resp, body, &userinfo); err != nil {
+		return nil, fmt.Errorf("failed to decode OIDC discovery document: %w", err)
+	}
+	return userinfo, nil
+}
+
+// validateAudience returns an error if audClaim does not contain any audiences
+// given by expectedAudiences. If expectedAudiences is empty, it skips validation
+// and returns nil.
+func validateAudience(expectedAudiences, audClaim []string) error {
+	if len(expectedAudiences) == 0 {
+		return nil
+	}
+
+	for _, v := range expectedAudiences {
+		if contains(audClaim, v) {
+			return nil
+		}
+	}
+
+	return errors.New("audience claim does not match any expected audience")
+}
+
+func contains(sl []string, st string) bool {
+	for _, s := range sl {
+		if s == st {
+			return true
+		}
+	}
+	return false
+}
+
+// createCAContext returns a context with a custom TLS client that's configured with the root
+// certificates from caPEM. If no certificates are configured, the original context is returned.
+func createCAContext(ctx context.Context, caPEM string) (context.Context, error) {
+	if caPEM == "" {
+		return ctx, nil
+	}
+
+	certPool := x509.NewCertPool()
```

Comment:
in future, i'll add cache for this cert poll.

Because this validator create every request.

Now, we only can add 'TODO' for this

Reply:
oh, my bad. This object caching here https://github.com/flant/negentropy/blob/main/vault-plugins/flant_iam_auth/usecase/authn/factory/factory.go